operators/pkg/controller/common/annotation/controller_version.go

```diff
@@ -0,0 +1,42 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package annotation
+
+import (
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// ControllerVersionAnnotation is the annotation name that indicates the last controller version to update a resource
+const ControllerVersionAnnotation = "k8s.elastic.co/controller-version"
+
+// UpdateControllerVersion updates the controller version annotation to the current version if necessary
+func UpdateControllerVersion(client k8s.Client, obj runtime.Object, version string) error {
+	metaObject, err := meta.Accessor(obj)
+	if err != nil {
+		log.Error(err, "error converting runtime object to metav1 object")
+		return err
+	}
+	annotations := metaObject.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// do not send extraneous update if the value would not change
+	if annotations[ControllerVersionAnnotation] == version {
+		return nil
+	}
+
+	annotations[ControllerVersionAnnotation] = version
+	accessor := meta.NewAccessor()
+	err = accessor.SetAnnotations(obj, annotations)
```

Comment:
Can we not reuse `metaObject.SetAnnotations` here, instead of creating a new accessor?

Reply:
I don't think so because accessor's `SetAnnotations(obj, annotations)` returns a `runtime.Object`, which is what we need for the `Update` call, while `metaObject.SetAnnotations()` modifies the metav1.Object receiver.

But your question did make me realize that I didn't look closely enough at the funcs available to the metav1 Accessor so there's some small bit of refactoring I can do here

Reply:
> metaObject.SetAnnotations() modifies the metav1.Object receiver

I think it also modifies the original object you passed to it, which you can the use for the update call.
But I'm also fine with the way you used it here :)

Reply:
Ah that was not my understanding just from scanning through the code but it's deeeeefinitely possible I am misunderstanding